### PR TITLE
✨(bbb-stress-test) initial working version of the stress testing tool

### PR DESCRIPTION
**Pull request by madmatah**
## Purpose

We want to launch a stress test on a BBB server from another server.

## Proposal

- [x] Initial version of the stress testing tool based on Puppeteer.
- [x] README with Getting started instructions
- [x] CircleCI config

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+# docker user (see entrypoint).
+RUN chmod g=u /etc/passwd
+
+# We wrap commands run in this container by the following entrypoint that
+# creates a user on-the-fly with the container user ID (see USER) and root group
+# ID.
+ENTRYPOINT [ "/usr/local/bin/entrypoint" ]
+
+# Un-privileged user running the application
+ARG DOCKER_USER
```
Comment (jmaupetit) [inline Dockerfile:28]: You don't provide a default?

```suggestion
ARG DOCKER_USER 1000
```

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+CMD ["google-chrome-unstable"]
+
+# ---- Development image ----
+
+FROM core as development
+
+CMD ["/bin/bash"]
+
+# ---- Image to publish ---- #
```
Comment (jmaupetit) [inline Dockerfile:39]:
```suggestion
# ---- Image to publish ----
```

File: Makefile [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+# -- Docker
+# Get the current user ID to use for docker run and docker exec commands
+DOCKER_UID           = $(shell id -u)
+DOCKER_GID           = $(shell id -g)
+DOCKER_USER          = $(DOCKER_UID):$(DOCKER_GID)
+
+COMPOSE              = DOCKER_USER=$(DOCKER_USER) docker-compose
+COMPOSE_RUN          = $(COMPOSE) run --rm
+COMPOSE_RUN_APP      = $(COMPOSE_RUN) app
+COMPOSE_RUN_NODE      = $(COMPOSE_RUN) -e HOME="/tmp" app
```
Comment (jmaupetit) [inline Makefile:35]:
```suggestion
COMPOSE_RUN_NODE     = $(COMPOSE_RUN) -e HOME="/tmp" app
```

File: Makefile [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+list-meetings: ## list meetings running on the BBB server
+	@$(COMPOSE_RUN_APP) ./cli.js list-meetings -v
+.PHONY: list-meetings
+
+
+# -- Project bootstrap
+
+.env:
+	@if [ ! -f .env ]; then cp .env.default .env ; echo ".env file generated successfully. Please edit it to set BBB_URL, BBB_SECRET and BBB_MEETING_ID" ; fi
+.PHONY: .env
```
Comment (jmaupetit) [inline Makefile:59]:
```suggestion
.env:
	cp .env.default .env
	@echo ".env file generated successfully. Please edit it to set BBB_URL, BBB_SECRET and BBB_MEETING_ID"
```

File: Makefile [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+
+bootstrap: ## Prepare Docker images for the project
+bootstrap: \
+	build \
+	.env
+.PHONY: bootstrap
+
+# -- Build tools
+
+build: ## build front-end application
```
Comment (jmaupetit) [inline Makefile:70]:
```suggestion
build: ## Build front-end application
```

File: Makefile [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+YARN                 = $(COMPOSE_RUN_NODE) yarn
+
+# ==============================================================================
+# RULES
+
+default: help
+
+# -- Test suite
+
+stress: ## run stress test
```
Comment (jmaupetit) [inline Makefile:46]:
```suggestion
stress: ## Run stress test
```

File: Makefile [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+	@$(COMPOSE_RUN_APP) ./cli.js stress -v
+.PHONY: stress
+
+list-meetings: ## list meetings running on the BBB server
```
Comment (jmaupetit) [inline Makefile:50]:
```suggestion
list-meetings: ## List meetings running on the BBB server
```

File: Makefile [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+build: \
+	build-image \
+	install
+.PHONY: build
+
+build-image: ## Build the docker image
+	docker-compose build app
+.PHONY: build-image
+
+install: ## install dependencies
```
Comment (jmaupetit) [inline Makefile:80]:
```suggestion
install: ## Install dependencies
```

File: Makefile [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+	@$(YARN) install
+.PHONY: install
+
+# -- Node
+
+lint: ## run all linters
```
Comment (jmaupetit) [inline Makefile:86]:
```suggestion
lint: ## Run linters
```

File: Makefile [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+lint: \
+  lint-prettier
+.PHONY: lint-front
```
Comment (jmaupetit) [inline Makefile:89]:
```suggestion
.PHONY: lint
```

File: Makefile [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+
+lint-prettier: ## run prettier over js/jsx/json/ts/tsx files -- beware! overwrites files
```
Comment (jmaupetit) [inline Makefile:91]:
```suggestion
lint-prettier: ## Run prettier over js/jsx/json/ts/tsx files -- beware! overwrites files
```

File: README.md [hunk 12]
```diff
@@ -4,6 +4,34 @@
...
+### Preparation
+
+1) Clone this repository
+
+2) Run `make bootstrap`
+
+3) Update the generated `.env` file to specify `BBB_URL` and `BBB_SECRET`. \
+You can get these values by running `bbb-conf --secret` on your BBB server.
+
+### Ready to launch your test ?
```
Comment (jmaupetit) [inline README.md:20]:
```suggestion
### Ready to launch your test?
```

File: README.md [hunk 13]
```diff
@@ -4,6 +4,34 @@
...
+
+1) Manually start a meeting on your BBB server.
+
+2) Get the meeting id by running `make list-meetings`
```
Comment (jmaupetit) [inline README.md:24]:
```suggestion
2) Get the meeting ID by running `make list-meetings`
```

File: cli.js [hunk 14]
```diff
@@ -0,0 +1,41 @@
+#!/usr/bin/env node
+
+const { getEnv } = require("./lib/command-helpers");
+
+const yargs = require("yargs");
+const _ = require("lodash/fp");
```
Comment (jmaupetit) [inline cli.js:6]: It does not seem to be used?

File: docker-compose.yml [hunk 15]
```diff
@@ -0,0 +1,16 @@
...
+
+services:
+  app:
+    build:
+      context: .
+      target: development
+    image: bbb-stress-test:development
+    working_dir: /app/
+    user: ${DOCKER_USER:-1000}
+    cap_add:
```
Comment (jmaupetit) [inline docker-compose.yml:11]: Didn't know about this :clap:

File: Makefile (hunk 4 above)
Comment (madmatah) [reply Makefile:59]: Oops, right, `make` can do that for us :+1:

File: docker-compose.yml (hunk 15 above)
Comment (madmatah) [reply docker-compose.yml:11]: Thanks to [this documentation](https://github.com/puppeteer/puppeteer/blob/main/docs/troubleshooting.md#running-puppeteer-in-docker) :slightly_smiling_face:

File: cli.js (hunk 14 above)
Comment (madmatah) [reply cli.js]: That's right ! :+1:

File: Makefile (hunk 4 above)
Comment (lunika) [inline Makefile]:
```suggestion

```

File: Makefile [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+
+list-meetings: ## list meetings running on the BBB server
+	@$(COMPOSE_RUN_APP) ./cli.js list-meetings -v
+.PHONY: list-meetings
+
+
+# -- Project bootstrap
+
+.env:
+	@if [ ! -f .env ]; then cp .env.default .env ; echo ".env file generated successfully. Please edit it to set BBB_URL, BBB_SECRET and BBB_MEETING_ID" ; fi
```
Comment (lunika) [inline Makefile]: Why this `if` statement ? Make will not exute this task if the file already exists

```suggestion
	cp .env.default .env
	@ echo ".env file generated successfully. Please edit it to set BBB_URL, BBB_SECRET and BBB_MEETING_ID" 
```

File: Makefile (hunk 10 above)
Comment (lunika) [inline Makefile]:
```suggestion
.PHONY: lint
```

File: Dockerfile (hunk 1 above)
Comment (madmatah) [reply Dockerfile]: I had to change a few things to make it work with a default value declared here.
Please look at af10a499458c82b875ea6b386434d2c6ca23adea

**Comment by madmatah** [comment]
Thanks @jmaupetit for your review :pray: 
I addressed all your feedbacks.

File: lib/commands/list-meetings.js [hunk 17]
```diff
@@ -0,0 +1,37 @@
+const BBB = require("../BBB");
+const { getLogger } = require("../command-helpers");
+const _ = require("lodash/fp");
+
+module.exports = {
+  name: "list-meetings",
+
+  description: "List meetings currently running on BBB",
+
+  options: (yargs) => yargs,
```
Comment (lunika) [inline lib/commands/list-meetings.js:10]: is it mandatory ? Because it looks like useless

File: lib/commands/list-meetings.js [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+        }
+        logger.info("Currently running meetings :");
+        for (const meeting of meetings) {
+          logger.info(
+            ` - ${meeting.meetingID} (${meeting.meetingName}) – participants : ${meeting.participantCount}`
+          );
+          const attendees = [].concat(_.get("attendees.attendee")(meeting));
+          if (attendees.length > 0) {
+            const attendeesNames = _.map("fullName")(attendees);
+            logger.debug(`   [${attendeesNames.join(", ")}]`);
```
Comment (lunika) [inline lib/commands/list-meetings.js:31]:
```suggestion
            logger.debug(`[${attendeesNames.join(", ")}]`);
```

File: lib/stress-test.js [hunk 19]
```diff
@@ -0,0 +1,122 @@
...
+    // If we are muted, click on Unmute
+    const unmuteButton = await page.$('[aria-label="Unmute"]');
+    if (unmuteButton !== null) {
+      logger.debug("Click on Unmute");
+      await unmuteButton.click();
+    }
+  }
+  if (webcam) {
+    await page.waitForSelector('[aria-label="Share webcam"]');
+    // await page.pdf({path: '/tmp/test1.pdf'});
```
Comment (lunika) [inline lib/stress-test.js:41]:
```suggestion

```

File: lib/stress-test.js [hunk 20]
```diff
@@ -0,0 +1,122 @@
...
+    await page.click('[aria-label="Share webcam"]');
+    logger.debug("clicked on sharing webcam");
+    await new Promise((resolve) => setTimeout(resolve, 2000));
+    // await page.pdf({path: '/app/screen.pdf'});
```
Comment (lunika) [inline lib/stress-test.js:45]:
```suggestion

```

File: lib/BBB.js [hunk 21]
```diff
@@ -0,0 +1,45 @@
...
+      return Promise.reject(`${response.messageKey} : ${response.message}`);
+    }
+    return Promise.resolve(response);
+  }
+
+  async getMeetings() {
+    return (
+      this.bbbApiQuery(this.api.monitoring.getMeetings())
+        .then(_.getOr([], "meetings.meeting"))
+        // The returned value can be a single element or an array, so we convert to an array in all cases
```
Comment (jmaupetit) [inline lib/BBB.js:23]:
```suggestion
        // The returned value can be a single element or an array, so we convert it to an array in all cases
```

File: lib/commands/list-meetings.js [hunk 22]
```diff
@@ -0,0 +1,37 @@
...
+    const bbbClient = new BBB(argv.u, argv.s);
+
+    bbbClient
+      .getMeetings()
+      .then((meetings) => {
+        if (meetings.length === 0) {
+          logger.info("No meeting running on the BBB server");
+          return;
+        }
+        logger.info("Currently running meetings :");
```
Comment (jmaupetit) [inline lib/commands/list-meetings.js:23]:
```suggestion
        logger.info("Currently running meetings:");
```

File: lib/username.js [hunk 23]
```diff
@@ -0,0 +1,50 @@
...
+  "Joanie",
+  "K.C.",
+  "Kirk",
+  "Leather",
+  "Lieutenant Colonel Binicky",
+  "Lieutenant Nurse Quinlan",
+  "Lori Beth",
+  "Louisa",
+  "Marion",
+  "Marsha",
```
Comment (lunika) [inline lib/username.js:27]: :heart:

File: lib/commands/stress.js [hunk 24]
```diff
@@ -0,0 +1,66 @@
...
+        alias: "d",
+        type: "number",
+        description:
+          "Duration of the stress test in seconds (after all clients are connected)",
+      })
+      .default("meeting", getEnv("BBB_MEETING_ID"))
+      .default("d", Number(getEnv("BBB_TEST_DURATION", 60)))
+      .default("w", Number(getEnv("BBB_CLIENTS_WEBCAM", 1)))
+      .default("m", Number(getEnv("BBB_CLIENTS_MIC", 1)))
+      .default("l", Number(getEnv("BBB_CLIENTS_LISTEN_ONLY", 2)))
```
Comment (jmaupetit) [inline lib/commands/stress.js:40]: Shouldn't default values get stored in variables like `DEFAULT_BBB_CLIENTS_LISTEN_ONLY`?

File: lib/stress-test.js [hunk 25]
```diff
@@ -0,0 +1,122 @@
...
+  await page.pdf({ path: "/tmp/test1.pdf" });
+  await page.waitForSelector(`[aria-label="${audioAction}"]`);
+  logger.debug(`click on ${audioAction}`);
+  await page.click(`[aria-label="${audioAction}"]`);
+  await page.waitForSelector(".ReactModal__Overlay", { hidden: true });
+  if (microphone) {
+    logger.debug("Ensure that we are not muted...");
+    // Wait for the toolbar to appear
+    await page.waitForSelector('[aria-label="Mute"],[aria-label="Unmute"]');
+    // In all cases, click on the mute button
```
Comment (jmaupetit) [inline lib/stress-test.js:25]:
```suggestion
    // In any case, click on the mute button
```

File: Makefile (hunk 16 above)
Comment (madmatah) [reply Makefile]: yes, I forgot that :smile:

File: lib/commands/list-meetings.js (hunk 17 above)
Comment (madmatah) [reply lib/commands/list-meetings.js:10]: Right, I should make it optional and handle that in `cli.js`.

File: lib/commands/list-meetings.js (hunk 18 above)
Comment (madmatah) [reply lib/commands/list-meetings.js:31]: I keep it like this to have a consistent list output : 

```
info:     - meeting-id1 (Meeting 1) – participants : 1
debug:      [Matthieu]
info:     - meeting-id2 (Home Room) – participants : 1
debug:      [Matthieu]
info:     - meeting-id3 (Meeting 3) – participants : 0
debug:      []
```

With your suggestion I would get : 
```
info:     - meeting-id1 (Meeting 1) – participants : 1
debug:   [Matthieu]
info:     - meeting-id2 (Home Room) – participants : 1
debug:   [Matthieu]
info:     - meeting-id3 (Meeting 3) – participants : 0
debug:   []
```

File: lib/stress-test.js (hunk 20 above)
Comment (madmatah) [reply lib/stress-test.js]: :man_facepalming:

**Comment by madmatah** [comment]
Thanks @lunika and @jmaupetit .
I pushed a fixup commit ( b38002df0b56813f3f0b1a821103517ca8ce77b2 ) :slightly_smiling_face:

**Comment by madmatah** [comment]
@lunika @jmaupetit I think it's ready to merge, may I ask you a final review :pray:  ?

**Pull request merged**